sstables/trie/trie_writer.hh

```diff
@@ -0,0 +1,1192 @@
+    sink_pos _pos;
+
+    static ptr<writer_node> create(const_bytes b, bump_allocator&);
+    // Emplaces a new child node into `_children`,
+    // and returns a pointer to it.
+    // Children must be added in order of rising transitions.
+    ptr<writer_node> add_child(const_bytes b, bump_allocator&);
+    // Accessors.
+    std::span<const ptr<writer_node>> get_children() const;
+    std::span<ptr<writer_node>> get_children();
+    // Resets the _children vector.
+    void reset_children();
+    // Ensures children has size at least n, growing it if needed.
+    void reserve_children(size_t n, bump_allocator& alctr);
+    // Appends the new child to _children. Expects sufficient _children_capacity reserved in advance.
+    void push_child(ptr<writer_node> x, bump_allocator& alctr);
+    // Returns the first character of this node's transition chain.
+    std::byte transition() const;
+    // Sets the payload.
+    //
+    // Must only be used on nodes immediately after they are created,
+    // i.e. before any other nodes are added.
+    //
+    // It's separate from `add_child` because the root might need a payload too,
+    // but it isn't created by `add_child`.
+    void set_payload(const trie_payload&) noexcept;
+    // Re-calculates _node_size and/or _branch_size of this node and its descendants
+    // (as needed by _has_out_of_page_children or _has_out_of_page_children) under
+    // the assumption that we will write out the entire subtree without any padding,
+    // with output initially at position start_pos. If the assumption is true,
+    // then the values of _node_size and _branch_size are exact after this call.
+    static sink_offset recalc_sizes(ptr<writer_node> self, const trie_writer_sink auto&, sink_pos start_pos);
+    // Writes out the entire subtree.
+    //
+    // When it's known in advance that the entire subtree will fit within the current page,
+    // guaranteed_fit can be set to true. This is the typical case.
+    //
+    // Otherwise, guaranteed_fit must be set to false. In this case it's legal for the subtree
+    // to not fit within current page (internal padding will be inserted accordingly so that
+    // each node doesn't cross node boundaries), but it will re-calculate sizes again,
+    // so it's more expensive.
+    //
+    // After this call, it's true that:
+    // get_children().empty() && _pos.valid() && !_has_out_of_page_children && !_has_out_of_page_descendants
+    static void write(ptr<writer_node> self, trie_writer_sink auto&, bool guaranteed_fit);
+};
+
+inline ptr<writer_node> writer_node::create(const_bytes b, bump_allocator& alctr) {
+    // We could do this in a single allocation, though we would have to change the interface
+    // of bump_allocator somewhat to enable that.
+    auto x = alctr.alloc<writer_node>();
+    new (x.get()) writer_node;
+    auto transition = alctr.alloc<std::byte[]>(b.size());
+    std::memcpy(transition.get(), b.data(), b.size());
```

Comment:
Limiting transition sizes and inlining them would improve performance (if you copy the whole thing unconditionally).

Clustering keys are limited to 64k and are typically much smaller. What about limiting transitions to 128 bytes?

Reply:
We can keep this in mind as an optimization idea, but this is something that has to be tested under a benchmark. I'm not convinced it would improve performance. For example, in a partition index, the transition will almost always be 1 byte. If you inline a 128 byte buffer, you are wasting memory. This could be worse than a call to memcpy.

Reply:
> Can you share your experience?

I just already played with this kind of optimizations (mostly by running `perf-sstable`) , and I did try eliminating some memcpy calls. And my conclusion was that, for the moment, this kind of optimization (i.e. making `memcpy` constant-sized) isn't worthwhile. 

The trie construction logic is heavy in general. It seems it's going to slow down things like flushes and easy compactions by about 5-10%.

(That's probably the main disadvantage of this format: the writer logic is fairly complicated and branchy, both for the programmer and for the CPU. In a simpler index, you could just append keys to a file and hoist one key every 4 kiB to a higher tree level).

Reply:
Seems okay to me and we can defer optimizations until later.